lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -282,38 +588,143 @@ public void OnCompleted()
         {
         }
 
-        private bool AreIMRUTasksCompleted()
+        private void UnexpectedState(string id, string eventName)
         {
-            return _completedTasks.Count >= _totalMappers + 1;
+            var msg = string.Format(CultureInfo.InvariantCulture,
+                "Received {0} for [{1}], but system status is {2}.",
+                eventName,
+                id,
+                _systemState.CurrentState);
+            Exceptions.Throw(new IMRUSystemException(msg), Logger);
         }
 
-        private string GetTaskIdByEvaluatorId(string evaluatorId)
+        /// <summary>
+        /// If all the tasks are in final state, if the system is recoverable, start recovery
+        /// else, change the system state to Fail then take Fail action
+        /// </summary>
+        private void TryRecovery()
+        {
+            if (_taskManager.AllInFinalState())
+            {
+                if (Recoverable())
+                {
+                    _isFirstTry = false;
+                    StartAction();
+                }
+                else
+                {
+                    Logger.Log(Level.Warning, "The system is not recoverable, change the state to Fail.");
+                    _systemState.MoveNext(SystemStateEvent.NotRecoverable);
+                    FailAction();
+                }
+            }
+        }
+
+        private string GetMapperTaskIdByEvaluatorId(string evaluatorId)
         {
             return string.Format("{0}-{1}-Version0",
                 IMRUConstants.MapTaskPrefix,
                 _serviceAndContextConfigurationProvider.GetPartitionIdByEvaluatorId(evaluatorId));
         }
 
+        private void DoneAction()
+        {
+            ShutDownAllEvaluators();
+        }
+
+        private void FailAction()
+        {
+            ShutDownAllEvaluators();
```

Comment:
So far, both dispose all the contexts. DoneAction is for successful scenario and FailAction is for failed and no recovery scenario.